src/Elmish.WPF/InternalUtils.fs

```diff
@@ -68,6 +68,29 @@ module ValueOption =
     | Ok x -> ValueSome x
     | Error _ -> ValueNone
 
+  [<RequireQualifiedAccess>]
+  type ToNullError =
+    | ValueCannotBeNull
+    | OtherExn of exn
+
+  let ofNull<'a> (a: 'a) =
+    match box a with
+    | null -> ValueNone
+    | _ -> ValueSome a
+
+  let toNull<'a> =
+    function
+    | ValueSome a -> Ok a
+    | ValueNone ->
+        let isNotNullable = typeof<'a>.IsValueType && System.Nullable.GetUnderlyingType(typeof<'a>) = null
+        if isNotNullable then
+          ToNullError.ValueCannotBeNull |> Error
+        else
+          try
+            null |> unbox<'a> |> Ok
+          with e ->
+            e |> ToNullError.OtherExn |> Error
+
```

Comment:
In a previous commit, the implementations were (sort of) duals of each other.  Now they are very different.

Why the new implementations?

`ofNull` seems simply enough.  But `toNull` seems less convincingly correct to me.  I recall a recent commit where `toNull` was nearly the same as it is now except that line 85 was different.  How do we know that 85 is correct now?  Maybe there is another condition missing.

Reply:
That's the whole problem. F# has no way of checking directly if a type is nullable or not at runtime. You can only do it at compile time. That's why I have to also use a try/catch block on the unbox to cover the cases that are incomplete.

Reply:
What is a specific case that was not handled correctly by your original implementation but is handled correctly by this implementation?

Reply:
Originally it would see `int?` as non-nullable, so I had to add an explicit check for the nullable type.

Reply:
I don't see any link in GitHub to the original implementation of `toNull`.  Can you share it here?

Reply:
Which one, the original one with boxing and unboxing, or the one that just did the `IsValueType` check?

Reply:
I meant the one with boxing/unboxing, but let's just put both of them here.

Reply:
```fsharp
let ofNull<'a> =
  box<'a>
  >> ValueOption.ofObj
  >> ValueOption.map unbox<'a>

let toNull<'a> =
  ValueOption.map box<'a>
  >> ValueOption.toObj
  >> unbox<'a>
```

Reply:
```fsharp
[<RequireQualifiedAccess>]
type ToNullError =
  | ValueCannotBeNull
  | OtherExn of exn

let ofNull<'a> (x: 'a) =
  match box x with
  | null -> ValueNone
  | _ -> ValueSome x

let toNull<'a> =
  function
  | ValueSome x ->
    x |> Ok
  | ValueNone ->
    if typeof<'a>.IsValueType then
      ToNullError.ValueCannotBeNull |> Error
    else
      try
        null |> unbox<'a> |> Ok
      with e ->
        e |> ToNullError.OtherExn |> Error
```

Reply:
The reason for the current implementation is to avoid throwing a `NullReferenceException` on the `unbox<'a>` call in the `toNull<'a>` function in cases where `ValueNone` is passed in and `'a` is not nullable.

Reply:
But in a sense, we're doing something that is no more dangerous than any other box/unbox situation.

Reply:
Or we could write it more like this:

```fsharp
  [<RequireQualifiedAccess>]
  type ToNullError =
    | ValueCannotBeNull

  let ofNull<'a> (x: 'a) =
    match box x with
    | null -> ValueNone
    | _ -> ValueSome x

  let toNull<'a> =
    function
    | ValueSome x ->
      x |> Ok
    | ValueNone ->
      if box Unchecked.defaultof<'a> <> null then
        ToNullError.ValueCannotBeNull |> Error
      else
        null |> unbox<'a> |> Ok
```

Reply:
I pushed the last version. I think it's much clearer that it is correct. `box defaultof<'a> <> null` seems reasonable that it is an analogue of `box x = null`

Reply:
Yes, that looks more convincingly correct.  I will take a close look when I get back from being out of town for the long holiday weekend.

Reply:
To be clear, `null |> unbox<'a>` throws a `NullReferenceException`, correct?

Reply:
I think it depends on `'a`.

Reply:
Thanks for sharing those previous implementations @marner2.  However, I didn't end up needing them because your most recent implementation is very convincingly correct. :D